### PR TITLE
doc: add KIC 2.6 and Kong 3.x controller guide

### PR DESCRIPTION
**Pull request by rainest**
### Summary
Adds documentation for KIC 2.6. consisting of metadata updates, compat table updates, and a new article for Kong 3.x upgrades.

### Reason
Part of https://github.com/Kong/kubernetes-ingress-controller/issues/2880

Normally I would submit docs after the KIC release is actually out, but in this case I'd like to link to the doc from the KIC changelog. I think we probably just allow the changelog link to be a dead link temporarily and merge this once the release is available, but wasn't sure we wanted to order it that way.

### Testing
Awaiting preview.

Partial runthrough of the new guide commands. I tested the commands to create the IngressClassParameters resource and patch the IngressClass. The rest are copied from https://docs.konghq.com/kubernetes-ingress-controller/latest/guides/upgrade/ with minor adjustments (updating versions and adding a few `--set` parameters for the KIC/Kong versions and I did  not test them separately. Chart 2.13 isn't actually out yet; the controller needs to be released before that will happen.

### CODE REVIEW

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+
+If you use a non-standard namespace and/or ingress class, you will need to
+replace the `kong` ingress class name and `kong` namespace with values
+appropriate to your environment.
+
+After you have upgraded to Kong 3.x, you should work to update all of your
+regular expression paths to include the `~` prefix in the Ingress itself at
+your earliest convenience. The heuristic is available for upgrade convenience
+and thus behaves exactly like the version included in Kong 2.8.x, bugs
+included. Once you added the prefix to all regular expression paths, you can
```
Comment (lena-larionova) [inline src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:139]:
```suggestion
After you have upgraded to Kong 3.x, update all of your regular expression paths 
to include the `~` prefix in the Ingress itself at your earliest convenience. 
The heuristic is only intended for decreasing downtime during upgrades, so 
it behaves exactly like the version included in Kong 2.8.x, bugs
included. Once you added the prefix to all regular expression paths, you can
```

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+disable `enableLegacyRegexDetection`.
+
+Lastly, Gateway API HTTPRoute resources are not affected by this problem: they
+do have a dedicated regular expression path type, and KIC does insert the `~`
+prefix automatically for these.
```
Comment (lena-larionova) [inline src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:144]:
```suggestion
Gateway API HTTPRoute resources are not affected by this problem. They
do have a dedicated regular expression path type, and KIC inserts the `~`
prefix automatically for these.
```

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+Using the existing chart version and the `values.yaml` configuration for
+your production environment, deploy a copy to your test cluster
+with the `--version` flag:
+
+```shell
+$ helm install kong-upgrade-testing kong/kong \
+  --version ${YOUR_VERSION} \
+  -f ${PATH_TO_YOUR_VALUES_FILE}
+```
+> **Note:** You may need to adjust your chart further to work in a development or
```
Comment (lena-larionova) [inline src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:184]:
```suggestion

{:.note}
> **Note:** You may need to adjust your chart further to work in a development or
```

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+you've configured {{site.base_gateway}}, and the existing `values.yaml` configuration file:
+
+```shell
+$ helm upgrade ${YOUR_RELEASE_NAME} kong/kong \
+  --namespace ${YOUR_NAMESPACE} \
+  -f ${PATH_TO_YOUR_VALUES_FILE} \
+  --set image.tag="3.0"
+```
+
+After the upgrade completes there is a brief period of time before the new
```
Comment (lena-larionova) [inline src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:260]:
```suggestion
After the upgrade completes, there is a brief period of time before the new
```

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+resources are online. You can wait for the relevant Pod resources to complete
+by watching them in your release namespace:
+
+```shell
+$ kubectl -n ${YOUR_RELEASE_NAMESPACE} get pods -w
+```
+
+Once the new pods are in a `Ready` state, the upgrade is complete. Update your
+values.yaml file to use the new Kong and ingress controller image versions to
```
Comment (lena-larionova) [inline src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:269]:
```suggestion
`values.yaml` file to use the new Kong and ingress controller image versions to
```

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+{% endnavtabs %}
+
+In the above example, `kong-2.13.0` is the currently deployed chart version.
+
+Using the existing chart version and the `values.yaml` configuration for
+your production environment, deploy a copy to your test cluster
+with the `--version` flag:
+
+```shell
+$ helm install kong-upgrade-testing kong/kong \
```
Comment (lena-larionova) [inline src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:180]:
```suggestion
helm install kong-upgrade-testing kong/kong \
```

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+kong    https://charts.konghq.com
+```
+
+{% endnavtab %}
+{% endnavtabs %}
+
+If the repository is not present, add it:
+
+```shell
+$ helm repo add kong https://charts.konghq.com
```
Comment (lena-larionova) [inline src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:224]:
```suggestion
helm repo add kong https://charts.konghq.com
```

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+{% endnavtab %}
+{% endnavtabs %}
+
+### Perform the upgrade
+
+Run the following command, specifying the old release name, the namespace where
+you've configured {{site.base_gateway}}, and the existing `values.yaml` configuration file:
+
+```shell
+$ helm upgrade ${YOUR_RELEASE_NAME} kong/kong \
```
Comment (lena-larionova) [inline src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:254]:
```suggestion
helm upgrade ${YOUR_RELEASE_NAME} kong/kong \
```

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+  -f ${PATH_TO_YOUR_VALUES_FILE} \
+  --set image.tag="3.0"
+```
+
+After the upgrade completes there is a brief period of time before the new
+resources are online. You can wait for the relevant Pod resources to complete
+by watching them in your release namespace:
+
+```shell
+$ kubectl -n ${YOUR_RELEASE_NAMESPACE} get pods -w
```
Comment (lena-larionova) [inline src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:265]:
```suggestion
kubectl -n ${YOUR_RELEASE_NAMESPACE} get pods -w
```

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+values.yaml file to use the new Kong and ingress controller image versions to
+continue using these through future upgrades.
+
+### Rollback
+
+If you run into problems during or after the upgrade, Helm provides a
+rollback mechanism to revert to a previous revision of the release:
+
+```shell
+$ helm rollback --namespace ${YOUR_RELEASE_NAMESPACE} ${YOUR_RELEASE_NAME}
```
Comment (lena-larionova) [inline src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:278]:
```suggestion
helm rollback --namespace ${YOUR_RELEASE_NAMESPACE} ${YOUR_RELEASE_NAME}
```

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+```
+
+You can wait for the rollback to complete by watching the relevant Pod
+resources:
+
+```shell
+$ kubectl -n ${YOUR_RELEASE_NAMESPACE} get pods -w
```
Comment (lena-larionova) [inline src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:285]:
```suggestion
kubectl -n ${YOUR_RELEASE_NAMESPACE} get pods -w
```

File: src/kubernetes-ingress-controller/references/version-compatibility.md [hunk 12]
```diff
@@ -11,65 +11,69 @@ those versions' documentation.
...
+| Kong Enterprise 2.6.x     | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> |
+| Kong Enterprise 2.7.x     | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> |
+| Kong Enterprise 2.8.x     | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> | <i class="fa fa-check"></i> |
+| Kong Enterprise 2.8.x     | <i class="fa fa-times"></i> | <i class="fa fa-times"></i> | <i class="fa fa-times"></i> | <i class="fa fa-times"></i> | <i class="fa fa-times"></i> | <i class="fa fa-times"></i> | <i class="fa fa-check"></i> |
```
Comment (lena-larionova) [inline src/kubernetes-ingress-controller/references/version-compatibility.md:44]:
```suggestion
| Kong Enterprise 3.0.x     | <i class="fa fa-times"></i> | <i class="fa fa-times"></i> | <i class="fa fa-times"></i> | <i class="fa fa-times"></i> | <i class="fa fa-times"></i> | <i class="fa fa-times"></i> | <i class="fa fa-check"></i> |
```

**Review by lena-larionova — DISMISSED** [review]

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+Ingress rule paths have no way to indicate that a path is a regular expression.
+The `ImplementationSpecific` path type can contain either regular expression or
+non-regular expression paths. If you have existing Ingresses with regular
+expression paths, those paths will no longer be routed correctly if you upgrade
+to 3.x without updating configuration.
+
+The preferred option for configuring regular expression Ingress rules for Kong
+3.x is to add the `~` prefix to Ingress rules directly. However, this is not
+compatible with Kong 2.x, and therefore cannot be easily done prior to the 3.x
+upgrade.
```
[the remark this answers is not in the document]
Comment (rainest) [reply src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:93]: I do want to stress that `enableLegacyRegexDetection` isn't intended to be a long-term solution and that this (adding the prefix yourself) is what you need to do eventually up front. Is there some better way to note that before the paragraph after ("After you have upgraded to Kong 3.x, update all of your regular expression paths...") about it?

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 14]
```diff
@@ -0,0 +1,294 @@
+---
+title: Upgrading to Kong 3.x
+---
+
+This guide walks through the steps needed to upgrade to 2.1.x and later
+versions from earlier versions, and covers changes from older versions to help
+operators evaluate whether they need to make changes to their configuration. It
+also covers creation of a testing environment to test the upgrade.
```
[the remark this answers is not in the document]
Comment (rainest) [reply src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md]: Yep.

**Comment by rainest** [comment]
Github is being silly about displaying things where I accepted the change suggestion, but https://github.com/Kong/docs.konghq.com/pull/4444/files/dcb80a40651e08df1474ec292db4ea4b8b9630c1#r969955115 still has open questions.

File: src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+using the same tools that deployed your production cluster, or use
+a local development cluster such as [minikube][minikube] or [kind][kind].
+
+Using Helm, check the deployed chart version:
+
+{% navtabs codeblock %}
+{% navtab Command %}
+
+```shell
+$ helm list -A
```
Comment (cloudjumpercat) [inline src/kubernetes-ingress-controller/guides/upgrade-kong-3x.md:151]:
```suggestion
helm list -A
```

**Review by cloudjumpercat — APPROVED** [review]
LGTM!

**Pull request merged**